File: study/ts/script/scene/view/shop/ShopViewItemCell2.ts
const {ccclass, property} = cc._decorator;

import CommonShopItemCell2 from '../../../ui/component/CommonShopItemCell2'
import ListViewCellBase from '../../../ui/ListViewCellBase';
import { handler } from '../../../utils/handler';

@ccclass
export default class ShopViewItemCell2 extends ListViewCellBase {

   @property({
       type: cc.Node,
       visible: true
   })
   _resourceNode: cc.Node = null;

   @property({
       type: CommonShopItemCell2,
       visible: true
   })
   _itemInfo1: CommonShopItemCell2 = null;

   @property({
       type: CommonShopItemCell2,
       visible: true
   })
   _itemInfo2: CommonShopItemCell2 = null;

   @property({
       type: CommonShopItemCell2,
       visible: true
   })
   _itemInfo3: CommonShopItemCell2 = null;

   @property({
       type: CommonShopItemCell2,
       visible: true
   })
   _itemInfo4: CommonShopItemCell2 = null;

   private TOTAL_COUNT:number = 4;

   private BACKGROUND_NORMAL:number = 0;
   private BACKGROUND_START:number = 1;
   private BACKGROUND_END:number = 2;

    _init(id){
        this.setIdx(id);
    }

    onInit() {
        var size = this._resourceNode.getContentSize();
        this.node.setContentSize(size.width, size.height);
    }
    updateUI(index, itemLine:any[], tabIndex) {
        this.node.name = ('ShopViewItemCell2' + (index + 1));
        for (var i = 1; i<=this.TOTAL_COUNT; i++) {
            var item = this['_itemInfo' + i] as CommonShopItemCell2;
            if (item) {
                item.node.active = (false);
            }
        }
        for (var i=1; i<=itemLine.length;i++) {
            var item1 = itemLine[i-1];
            var itemInfo = this['_itemInfo' + i] as CommonShopItemCell2;
            if (itemInfo) {
                var bgStatus = this.BACKGROUND_NORMAL;
                if (i == 1) {
                    bgStatus = this.BACKGROUND_START;
                }
                if (i == this.TOTAL_COUNT) {
                    bgStatus = this.BACKGROUND_END;
                }
                itemInfo.updateUI(item1, false,tabIndex);
                itemInfo.node.active = (true);
                itemInfo.updateBackgroundImge(bgStatus);
                itemInfo.setItemID(i + index * this.TOTAL_COUNT);
                itemInfo.setCallBack(handler(this, this._onBtnClick));
            }
        }
    }
    _onBtnClick(iconNameNode:CommonShopItemCell2) {
        //dump(iconNameNode);
        var curSelectedPos = iconNameNode.getItemID();
        //logWarn(' ShopViewItemCell2:_onBtnClick  ' + curSelectedPos);
        this.dispatchCustomCallback(curSelectedPos);
    }
    findItemByPos(pos) {
        for (var i = 1; i<=this.TOTAL_COUNT; i++) {
            var itemInfo = this['_itemInfo' + i] as CommonShopItemCell2;
            if (itemInfo && itemInfo.getItemID() == pos) {
                return itemInfo;
            }
        }
        return null;
    }
}

File: study/lua/src/app/scene/view/shop/ShopViewItemCell2.lua
-- Author: zhanglinsen
-- Date:2018-08-08 12:50:24
-- Describle：
local ListViewCellBase = require("app.ui.ListViewCellBase")
local ShopViewItemCell2 = class("ShopViewItemCell2", ListViewCellBase)
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local TOTAL_COUNT = 4
local BACKGROUND_NORMAL = 0 --正常显示背景
local BACKGROUND_START = 1 -- 开始时显示背景
local BACKGROUND_END = 2 --结束时显示背景
function ShopViewItemCell2:ctor()
	self._itemInfo1 = nil
	self._itemInfo2 = nil
	self._itemInfo3 = nil
	self._itemInfo4 = nil

	local resource = {
		file = Path.getCSB("ShopViewItemCell2", "shop"),

	}
	ShopViewItemCell2.super.ctor(self, resource)
end

function ShopViewItemCell2:onCreate()
	local size = self._resourceNode:getContentSize()
	self:setContentSize(size.width, size.height)

	for i=1, TOTAL_COUNT do
		local item = self["_itemInfo"..i]
		if item then
			item:setVisible(false)
		end
	end
end

function ShopViewItemCell2:update(index, itemLine, tabIndex)
	--dump(itemLine)
	self:setName("ShopViewItemCell2"..index+1)
	for i=1, TOTAL_COUNT do
		local item = self["_itemInfo"..i]
		if item then
			item:setVisible(false)
		end
	end

	for i , item in pairs(itemLine) do
		local itemInfo = self["_itemInfo"..i]
		if itemInfo then
			local bgStatus = BACKGROUND_NORMAL;
			if i == 1 then bgStatus = BACKGROUND_START end
			if i == TOTAL_COUNT then bgStatus = BACKGROUND_END end
			itemInfo:updateUI(item, nil, tabIndex)
			itemInfo:setVisible(true)
			itemInfo:updateBackgroundImge(bgStatus)
			itemInfo:setTag(i + index*TOTAL_COUNT)
			itemInfo:setCallBack(handler(self,self._onBtnClick))
		end
	end
end


function ShopViewItemCell2:_onBtnClick(iconNameNode)
	dump(iconNameNode)
	local curSelectedPos = iconNameNode:getTag()
	logWarn(" ShopViewItemCell2:_onBtnClick  "..curSelectedPos)
	
	self:dispatchCustomCallback(curSelectedPos)
end


function ShopViewItemCell2:findItemByPos(pos)
	for i=1 , TOTAL_COUNT  do
		local itemInfo = self["_itemInfo"..i]
		if itemInfo and itemInfo:getTag() == pos then
			return itemInfo
		end
	end
	return nil
end


return ShopViewItemCell2